stackslib/src/chainstate/coordinator/mod.rs

```diff
@@ -293,33 +307,54 @@ impl RewardSetProvider for OnChainRewardSetProvider {
     ) -> Result<RewardSet, Error> {
         let cur_epoch = SortitionDB::get_stacks_epoch(sortdb.conn(), cycle_start_burn_height)?
             .expect(&format!(
-                "FATAL: no epoch for burn height {}",
-                cycle_start_burn_height
+                "FATAL: no epoch for burn height {cycle_start_burn_height}",
             ));
-        if cur_epoch.epoch_id < StacksEpochId::Epoch30 {
+        let cycle = burnchain
+            .block_height_to_reward_cycle(cycle_start_burn_height)
+            .expect("FATAL: no reward cycle for burn height");
+        let is_nakamoto_reward_set = match SortitionDB::get_stacks_epoch_by_epoch_id(
```

Comment:
Yeah it seems like this code contradicts a line from the PR description:
> The first nakamoto reward set will occur in epoch 2.5.

Reply:
> Should we be checking at epoch 2.5? Are all epoch 2.5 reward sets Nakamoto reward sets?

I don't think that they must be or will be. There's a couple reasons for this. First, there's basically two differences between nakamoto reward sets and 2.x reward sets: (1) they _must_ contain signing keys, and so the code asserts as much, and (2) if there's no participation, an error is yielded and burnchain processing halts.

For the first constraint, you can see that there will always be at least one reward cycle in 2.5 which didn't use pox-4. Imagine reward cycle 10 starts at block height 300, and epoch 2.5 starts at block height 300, then reward cycle 10 would be in epoch-2.5, but there's no way it could be a nakamoto reward set: it cannot load signing keys because it could not have been constructed in pox-4.

The second difference is even more important, though, because non-participation during 2.5 _doesn't_ mean a chain halts. It only means a chain halts if the reward-cycle is also part of epoch-3.0. This could be true of the last cycle in 2.5 (but no other ones).

> Yeah it seems like this code contradicts a line from the PR description:
>> The first nakamoto reward set will occur in epoch 2.5.

I think what's happening here is a little confusing -- let's say Epoch-3.0 starts at height 305 and each reward cycle is length 10. Reward cycle 30 is the first nakamoto reward set, even though its start height is in Epoch 2.5. What's important is that every reward set which is active at any time in Epoch 3.0 be a nakamoto reward set. So that's what this calculation is doing: figure out what cycle is active at epoch-3's start height and then find the cycle start height of that cycle.

Reply:
Great, thanks!